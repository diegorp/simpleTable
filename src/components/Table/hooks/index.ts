export { useTableColumns } from './useTableColumns';
export { useTableData } from './useTableData';
export { useTableSort } from './useTableSort';
export { useTableFilter } from './useTableFilter';
export { useTablePagination } from './useTablePagination';
