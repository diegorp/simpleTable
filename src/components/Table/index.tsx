import type { Sort, Filter } from '@/types';
import TableFilter from './components/TableFilter';
import TableHead from './components/TableHead';
import TableBody from './components/TableBody';
import Pagination from './components/Pagination';
import {
  useTableColumns,
  useTableData,
  useTableSort,
  useTableFilter,
  useTablePagination,
} from './hooks';

interface Props<T> {
  title?: string;
  data: T[];
  rowIdKey: keyof T;
  hideRowId?: boolean;
  pagination?: boolean;
  sort?: boolean;
  filter?: boolean;
  onPageChange?: (page: number) => void;
  onPageSizeChange?: (pageSize: number) => void;
  onSortChange?: (sort: Sort<T>) => void;
  onFilterChange?: (filter: Filter<T>) => void;
}

export default function Table<T>({
  title,
  data,
  rowIdKey,
  hideRowId = false,
  pagination,
  sort,
  filter,
  onPageChange,
  onPageSizeChange,
  onSortChange,
  onFilterChange,
}: Props<T>) {
  const columns = useTableColumns(data, rowIdKey, hideRowId);
  const { currentSort, toggleSort } = useTableSort(rowIdKey, onSortChange);
  const { activeFilter, applyFilter } = useTableFilter(onFilterChange);
  const { currentPage, currentPageSize, changePage, changePageSize } =
    useTablePagination(onPageChange, onPageSizeChange);
  const { rows, isLastPage, totalItems } = useTableData(
    data,
    currentPage,
    currentPageSize,
    currentSort,
    activeFilter
  );

  const updateFilter = (filter: Filter<T>) => {
    applyFilter(filter);
    changePage(0);
  };

  const shouldShowHeader = title || filter;

  return (
    <article className='flex flex-col gap-4 overflow-hidden'>
      {shouldShowHeader && (
        <header className='flex items-center justify-between'>
          {title && <h2 className='text-[15px]'>{title}</h2>}
          {filter && (
            <TableFilter
              initialFilter={activeFilter}
              filterCategories={columns}
              onFilterChange={updateFilter}
            />
          )}
        </header>
      )}
      <div className='flex-1 overflow-auto border border-[#E2E8F0]'>
        <table className='min-w-full table-fixed overflow-auto text-gray-800'>
          <TableHead
            columns={columns}
            sort={sort}
            currentSort={currentSort}
            onToggleSort={toggleSort}
          />
          <TableBody rows={rows} columns={columns} rowIdKey={rowIdKey} />
        </table>
      </div>
      {pagination && (
        <Pagination
          currentPage={currentPage}
          currentPageSize={currentPageSize}
          totalItems={totalItems}
          isLastPage={isLastPage}
          onPageChange={changePage}
          onPageSizeChange={changePageSize}
        />
      )}
    </article>
  );
}
