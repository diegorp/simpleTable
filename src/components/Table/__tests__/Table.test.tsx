import { render, fireEvent } from '@testing-library/react';
import Table from '@/components/Table';

beforeEach(() => {
  HTMLElement.prototype.hidePopover = vi.fn();
  HTMLElement.prototype.showPopover = vi.fn();
});

describe('<Table />', () => {
  const mockData = [
    { id: 1, name: 'Alice', age: 25 },
    { id: 2, name: 'Bob', age: 30 },
    { id: 3, name: 'Charlie', age: 35 },
  ];

  const defaultProps = {
    data: mockData,
    rowIdKey: 'id' as const,
  };

  describe('Rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<Table {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });

    it('should render table with data', () => {
      const { getByText } = render(<Table {...defaultProps} />);

      expect(getByText('Alice')).toBeInTheDocument();
      expect(getByText('Bob')).toBeInTheDocument();
      expect(getByText('Charlie')).toBeInTheDocument();
    });

    it('should display empty state when no data provided', () => {
      const { getByText } = render(<Table data={[]} rowIdKey="id" />);

      expect(getByText('No data to display.')).toBeInTheDocument();
    });

    it('should handle data updates through rerender', () => {
      const { rerender, getByText, queryByText } = render(
        <Table {...defaultProps} />
      );

      expect(getByText('Alice')).toBeInTheDocument();

      const newData = [
        { id: 4, name: 'David', age: 40 },
        { id: 5, name: 'Eve', age: 45 },
      ];

      rerender(<Table data={newData} rowIdKey='id' />);

      expect(queryByText('Alice')).not.toBeInTheDocument();
      expect(getByText('David')).toBeInTheDocument();
      expect(getByText('Eve')).toBeInTheDocument();
    });

    it('should render title when provided', () => {
      const { getByText } = render(<Table {...defaultProps} title="Test Table" />);

      expect(getByText('Test Table')).toBeInTheDocument();
    });

    it('should not render header when no title and no filter', () => {
      const { container } = render(<Table {...defaultProps} />);

      const header = container.querySelector('header');
      expect(header).not.toBeInTheDocument();
    });

    it('should render header when title is provided', () => {
      const { container } = render(<Table {...defaultProps} title="Test" />);

      const header = container.querySelector('header');
      expect(header).toBeInTheDocument();
    });

    it('should render header when filter is enabled', () => {
      const { container } = render(<Table {...defaultProps} filter={true} />);

      const header = container.querySelector('header');
      expect(header).toBeInTheDocument();
    });
  });

  describe('Column Configuration', () => {
    it('should hide row ID column when hideRowId is true', () => {
      const { queryByText } = render(<Table {...defaultProps} hideRowId={true} />);

      expect(queryByText('id')).not.toBeInTheDocument();
    });

    it('should show row ID column when hideRowId is false', () => {
      const { getByText } = render(<Table {...defaultProps} hideRowId={false} />);

      expect(getByText('id')).toBeInTheDocument();
    });
  });

  describe('Pagination Integration', () => {
    it('should render Pagination component when pagination is enabled', () => {
      const { getByText } = render(<Table {...defaultProps} pagination={true} />);

      expect(getByText('Rows per page:')).toBeInTheDocument();
    });

    it('should not render Pagination component when pagination is disabled', () => {
      const { queryByText } = render(<Table {...defaultProps} pagination={false} />);

      expect(queryByText('Rows per page:')).not.toBeInTheDocument();
    });
  });

  describe('Sort Integration', () => {
    it('should render sortable TableHead when sort is enabled', () => {
      const { getByRole } = render(
        <Table {...defaultProps} sort={true} hideRowId={true} />
      );

      const nameHeader = getByRole('button', { name: 'Sort by name' });
      expect(nameHeader).toBeInTheDocument();
    });

    it('should render non-sortable TableHead when sort is disabled', () => {
      const { queryByRole } = render(
        <Table {...defaultProps} sort={false} hideRowId={true} />
      );

      const nameHeader = queryByRole('button', { name: 'Sort by name' });
      expect(nameHeader).not.toBeInTheDocument();
    });
  });

  describe('Filter Integration', () => {
    it('should render TableFilter component when filter is enabled', () => {
      const { getByRole } = render(<Table {...defaultProps} filter={true} />);

      expect(getByRole('button', { name: 'Filter' })).toBeInTheDocument();
    });

    it('should not render TableFilter component when filter is disabled', () => {
      const { queryByRole } = render(<Table {...defaultProps} filter={false} />);

      expect(queryByRole('button', { name: 'Filter' })).not.toBeInTheDocument();
    });
  });

  describe('Callback Integration', () => {
    it('should call onPageChange callback when pagination changes', () => {
      const onPageChange = vi.fn();
      const paginatedData = [
        { id: 1, name: 'Alice', age: 25 },
        { id: 2, name: 'Bob', age: 30 },
        { id: 3, name: 'Charlie', age: 35 },
        { id: 4, name: 'David', age: 40 },
        { id: 5, name: 'Eve', age: 45 },
        { id: 6, name: 'Frank', age: 50 },
      ];
      const { getByRole } = render(
        <Table data={paginatedData} rowIdKey="id" pagination={true} onPageChange={onPageChange} />
      );

      const nextButton = getByRole('button', { name: 'Next page' });
      fireEvent.click(nextButton);

      expect(onPageChange).toHaveBeenCalledWith(1);
    });

    it('should call onPageSizeChange callback when page size changes', () => {
      const onPageSizeChange = vi.fn();
      const { getByRole } = render(
        <Table
          {...defaultProps}
          pagination={true}
          onPageSizeChange={onPageSizeChange}
        />
      );

      const select = getByRole('combobox', {
        name: 'Rows per page',
      });
      fireEvent.change(select, { target: { value: '25' } });

      expect(onPageSizeChange).toHaveBeenCalledWith(25);
    });

    it('should call onSortChange callback when sort changes', () => {
      const onSortChange = vi.fn();
      const { getByRole } = render(
        <Table
          {...defaultProps}
          sort={true}
          hideRowId={true}
          onSortChange={onSortChange}
        />
      );

      const nameHeader = getByRole('button', { name: 'Sort by name' });
      fireEvent.click(nameHeader);

      expect(onSortChange).toHaveBeenCalledWith({
        key: 'name',
        direction: 'asc',
      });
    });

    it('should call onFilterChange callback when filter is applied', () => {
      const onFilterChange = vi.fn();
      const { getByRole, getByPlaceholderText, getByLabelText, getByText } =
        render(
          <Table
            {...defaultProps}
            filter={true}
            onFilterChange={onFilterChange}
          />
        );

      const filterButton = getByRole('button', { name: 'Filter' });
      fireEvent.click(filterButton);

      const select = getByLabelText('Select a category for filtering');
      fireEvent.change(select, { target: { value: 'name' } });

      const input = getByPlaceholderText('Filter value...');
      fireEvent.change(input, { target: { value: 'Alice' } });

      const applyButton = getByText('Apply');
      fireEvent.click(applyButton);

      expect(onFilterChange).toHaveBeenCalledWith({
        key: 'name',
        value: 'Alice',
      });
    });
  });
});
