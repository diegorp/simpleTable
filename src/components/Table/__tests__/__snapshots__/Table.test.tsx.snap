// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Table /> > Rendering > should match snapshot 1`] = `
<div>
  <article
    class="flex flex-col gap-4 overflow-hidden"
  >
    <div
      class="flex-1 overflow-auto border border-[#E2E8F0]"
    >
      <table
        class="min-w-full table-fixed overflow-auto text-gray-800"
      >
        <thead>
          <tr>
            <th
              aria-sort="other"
              class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
              scope="col"
            >
              id
            </th>
            <th
              aria-sort="other"
              class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
              scope="col"
            >
              name
            </th>
            <th
              aria-sort="other"
              class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
              scope="col"
            >
              age
            </th>
          </tr>
        </thead>
        <tbody
          class="bg-white"
        >
          <tr
            class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
          >
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              1
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              Alice
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              25
            </td>
          </tr>
          <tr
            class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
          >
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              2
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              Bob
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              30
            </td>
          </tr>
          <tr
            class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
          >
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              3
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              Charlie
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
            >
              35
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </article>
</div>
`;
